nexus/db-queries/src/db/datastore/instance.rs

```diff
@@ -434,6 +479,55 @@ impl DataStore {
             .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))
     }
 
+    /// List all instances in the [`Failed`](InstanceState::Failed) state with an
+    /// auto-restart policy that permits them to be automatically restarted by
+    /// the control plane.
+    ///
+    /// This is used by the `instance_reincarnation` RPW to ensure that that any
+    /// such instances are restarted.
+    ///
+    /// This query returns `n` randomly-ordered instances which are eligible for
+    /// reincarnation. Because reincarnating an instance changes its state so
+    /// that it no longer matches this query, it isn't necessary to use
+    /// pagination to avoid the query returning the same instance multiple
+    /// times: instead, we just actually reincarnate it to remove it from the
+    /// result set. Randomizing the order in which instances are returned allows
+    /// a nicer distribution of work across multiple Nexus replicas'
```

Comment:
I'm a little worried that randomizing the order makes it impossible to use the index for this.  Can you check with `EXPLAIN` that we're not doing some really expensive query?

Reply:
Hm, it does look like the current query is kinda costly --- it only uses the `lookup_instance_by_project` index. I added a test that runs `EXPLAIN` for this (in https://github.com/oxidecomputer/omicron/pull/6503/commits/a0256e0d5201e22ddf918e0aa31ead5c4f3615c2), and it suggests creating some additional indices for this query: 
```
distribution: local
vectorized: true

• top-k
│ order: +column25
│ k: 16
│
└── • render
    │
    └── • filter
        │ filter: (((state = 'failed') AND (auto_restart_policy = 'best_effort')) AND (((time_last_auto_restarted IS NULL) OR ((auto_restart_cooldown IS NOT NULL) AND (time_last_auto_restarted <= ('2024-09-18 20:35:03.19219+00:00' - auto_restart_cooldown)))) OR ((auto_restart_cooldown IS NULL) AND (time_last_auto_restarted <= '2024-09-18 19:35:03.19219+00:00')))) AND (updater_id IS NULL)
        │
        └── • index join
            │ table: instance@instance_pkey
            │
            └── • scan
                  missing stats
                  table: instance@lookup_instance_by_project (partial index)
                  spans: FULL SCAN

index recommendations: 2
1. type: index creation
   SQL command: CREATE INDEX ON instance (state) STORING (name, description, time_created, time_modified, time_deleted, project_id, user_data, time_state_updated, state_generation, active_propolis_id, target_propolis_id, migration_id, ncpus, memory, hostname, updater_id, updater_gen, time_last_auto_restarted, auto_restart_policy, auto_restart_cooldown);
2. type: index creation
   SQL command: CREATE INDEX ON instance (auto_restart_policy) STORING (name, description, time_created, time_modified, time_deleted, project_id, user_data, time_state_updated, state_generation, active_propolis_id, target_propolis_id, migration_id, ncpus, memory, hostname, updater_id, updater_gen, state, time_last_auto_restarted, auto_restart_cooldown);
```

I'm happy to add indices for this --- I'm somewhat surprised we don't already have an instances by state index, as I think there are actually several other queries that could benefit from that.

@gjcolombo and I were discussing whether the randomized approach was actually the right thing or not this morning, as figuring out the best way to ensure instances don't appear in it multiple times within the same activation is kind of hairy. We may end up changing this back to be paginated in order, and accept the cost of duplicate start sagas (which will fail) when multiple Nexuses try to reincarnate a bunch of the same instances. 

In either case though, we probably ought to at least have an instance-by-state index...